tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * <p>Use {@code NameScope} to group related operations within a hierarchy, which for example lets
+ * tensorboard coalesce nodes for better graph visualizations.
+ *
+ * <p>This class is package private, user code creates {@link Scope} which internally delegates
+ * calls to an underlying {@code NameScope}.
+ *
+ * <p>This class is thread-safe.
+ */
+final class NameScope {
+
+  NameScope withSubScope(String scopeName) {
+    if (baseName == null) {
+      checkPattern(OP_NAME_REGEX, scopeName);
+    } else {
+      checkPattern(SUBSCOPE_NAME_REGEX, scopeName);
+    }
+
+    if (baseOpName != null) {
+      // Use the base name instead to derive the subscope.
+      scopeName = baseOpName;
+    }
+
+    String newBaseName = fullyQualify(makeUnique(scopeName));
+    return new NameScope(newBaseName, null, null);
+  }
+
+  NameScope withName(String name) {
+    checkPattern(OP_NAME_REGEX, name);
+
+    // All context except for the baseOpName is shared with the new scope.
+    return new NameScope(baseName, name, ids);
+  }
+
+  String makeOpName(String opName) {
+    checkPattern(OP_NAME_REGEX, opName);
+
+    if (baseOpName != null) {
+      opName = baseOpName;
+    }
+    opName = makeUnique(opName);
+    return fullyQualify(opName);
+  }
+
+  /**
+   * Create a new, root-level namescope.
+   *
+   * <p>A root-level namescope generates operator names with no components, like {@code Const_72}
+   * and {@code result}.
+   */
+  NameScope() {
+    this(null, null, null);
+  }
+
+  private NameScope(String baseName, String baseOpName, Map<String, Integer> ids) {
+    this.baseName = baseName;
+    this.baseOpName = baseOpName;
+    if (ids != null) {
+      this.ids = ids;
+    } else {
+      this.ids = new HashMap<String, Integer>();
+    }
+  }
+
+  // Generate a unique name, different from existing ids.
+  //
+  // ids is a map from id to integer, representing a counter of the
+  // number of previous requests to generate a unique name for the
+  // given id.
+  //
+  // For instance, the first use of makeUnique("a") adds "a" -> 1
+  // to ids and returns "a".
+  //
+  // The second use of makeUnique("a") updates ids to "a" -> 2
+  // and returns "a_1", and so on.
+  private String makeUnique(String id) {
+    synchronized (ids) {
+      if (!ids.containsKey(id)) {
+        ids.put(id, 1);
+        return id;
+      } else {
+        int cur = ids.get(id);
+        ids.put(id, cur + 1);
+        return String.format("%s_%d", id, cur);
+      }
+    }
+  }
+
+  private String fullyQualify(String name) {
+    if (baseName != null) {
+      return String.format("%s/%s", baseName, name);
+    } else {
+      return name;
+    }
+  }
+
+  // If baseName is non-null, it is a prefix applied to all names
+  // created by this instance.
+  private final String baseName;
+
+  // If baseOpName is non-null, it is used to derive the unique name
+  // for operators rather than the provided default name.
+  private final String baseOpName;
```

Comment:
How about `scopeName` and `opName` instead of `baseName` and `baseOpName`? Or perhaps `opPrefix` and `opName` - since one is intended to be a prefix for any ops created?